### PR TITLE
Adds a DediProg flash runner for board mec15xxevb_assy6853

### PR DESCRIPTION
**Pull request by dcpleung**
This adds a generic DediProg flash runner to be used with flashing the SPI chip for board mec15xxevb_assy6853.

### CODE REVIEW

**Review by mbolivar — APPROVED** [review]
Looks like a good start. If you can add some test cases and error handling for missing --spi-image that'd be great.

**Review by albertofloyd — APPROVED** [review]

**Comment by dcpleung** [comment]
> Looks like a good start. If you can add some test cases and error handling for missing --spi-image that'd be great.

I added "required=True" for both arguments. The ArgumentParser will print an error message if they are missing.

**Comment by dcpleung** [comment]
Add a few test cases.

**Review by mbolivar — APPROVED** [review]
Looks good. Thanks.

File: scripts/west_commands/runners/dediprog.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+    @classmethod
+    def create(cls, cfg, args):
+        return DediProgBinaryRunner(cfg,
+                                    spi_image=args.spi_image,
+                                    vcc=args.vcc)
+
+    def do_run(self, command, **kwargs):
+        self.require(DPCMD_EXE)
+        cmd_flash = [DPCMD_EXE, '-u', self.spi_image]
```
Comment (marc-hb) [inline scripts/west_commands/runners/dediprog.py]: Short options like `-u` are great to save time typing. In a script please prefer the longer and clearer version, here: `--auto`

**Review by marc-hb — DISMISSED** [review]
How extensively has this been tested? I don't know that particular board/system but my extensive experience with SF100 and SF600 on a number of other systems taught me a number of Dediprog lessons: `dpcmd --auto` is a good start but nowhere near enough. Here's what I think is missing:

- `--silent` otherwise the progress bar / counter generates about 2 gigabytes of extra logs in automation. Very easy to reproduce, just: `dpcmd ... |& tee big.log`
- `--verify` otherwise success is frequently claimed when not, about 20% of the time in my experience.
- a retry loop because of previous item.

For reference, here are a shell and a Powershell script that have been successfully tested across multiple platforms for long periods of time. Of course you may want to replace the infinite `while` loop with a limited `for` loop.

```
dpcmdautoloop() { while ! time ~/SF100Linux/dpcmd --auto "$1" --silent --verify ; do sleep 2; done;}
dpcmdautoloop ./spi_image.bin
```

```
function dpcmdautoloop($ifwi) {
    do { Measure-Command {
        & "C:\Program Files (x86)\DediProg\SF100\Dpcmd.exe" --auto $ifwi --silent --verify | Out-Default
    } ; start-sleep -s 2;
    } while ($lastexitcode)
}
```

**Comment by marc-hb** [comment]
> Of course you may want to replace the infinite while loop with a limited for loop.

Logging `--verify` failures and the number of attempts would also be extremely useful for CI: it would provide a great quality indicator even if the number of attempts is always high enough.

**Comment by dcpleung** [comment]
I have used `dpcmd` on this board possibly 100+ times without any issues. However, if you ever ctrl-c a running dpcmd, the programmer is put into a weird state which requires power cycling. I have never been able to "re-try" my way out of this situation.

But let me add the silent option. I can add a retry there but I don't know if I can test this part correctly.

**Comment by marc-hb** [comment]
> I have used dpcmd on this board possibly 100+ times without any issues.

That's probably because the EC image / delta with the existing image is very small.

`--auto` is incremental: it reprograms only what's needed. One thing I noticed when using the retry loops I shared above is some sort of "convergence" effect: as the delta becomes smaller and smaller after each attempt, the failure risk seemed to become smaller and smaller. In other words it seems the more you program the more likely there will be failure.

BTW I just tested (old) version https://github.com/DediProgSW/SF100Linux/commits/40bb62e5f807 and confirmed `--auto` did *not* imply `--verify` at the time.

> However, if you ever ctrl-c a running dpcmd, the programmer is put into a weird state which requires power cycling. I have never been able to "re-try" my way out of this situation.

Agreed: you likely want to be next to the board if you Ctrl-C. Not a problem for automation.

>  I can add a retry there but I don't know if I can test this part correctly.

If you try erasing first maybe you'll be more likely to see a failure. You can also just Ctrl-C and let it fail after the max of say 5 times: that will not test all situations but most of the retry code anyway.

File: boards/arm/mec15xxevb_assy6853/doc/index.rst [hunk 2]
```diff
@@ -274,6 +274,17 @@ Flashing
 
    .. note:: Remember that SPI MISO/MOSI are swapped on dediprog headers!
 
+   - It is possible to flash using west by doing:
```
Comment (dbkinder) [inline boards/arm/mec15xxevb_assy6853/doc/index.rst]: A more active way to say this is:
```
Flash your board using west:
```

File: boards/arm/mec15xxevb_assy6853/doc/index.rst [hunk 3]
```diff
@@ -274,6 +274,17 @@ Flashing
...
+
+     .. code-block:: console
+
+        $ west flash
+
+     Please make sure that the program ``dpcmd`` (on Linux) or
```
Comment (dbkinder) [inline boards/arm/mec15xxevb_assy6853/doc/index.rst]: While polite, we don't use ``please`` in our docs:
```
Make sure the program ``dpcmd`` ...
```

**Review by dbkinder — CHANGES_REQUESTED** [review]
A couple of small tweaks to the doc...

**Comment by dcpleung** [comment]
Added the silent and verify options, and updated the doc according to comments.

**Review by dbkinder — APPROVED** [review]
docs LGTM, thanks.

**Comment by dcpleung** [comment]
Updated the dediprog script with retries.

**Review by nashif — APPROVED** [review]

**Comment by maksimmasalski** [comment]
@dcpleung I can't use sanitycheck for programming of the Microchip board, it always gives me timeout error. I can't find out a root cause, but pretty sure that it is problem with dediprog runner.
If to run $west flash for the Microchip board separately, it works well. But if to run sanitycheck for the Microchip board, I always receive error "Chip can not be detected".
At the same time if to run command $west flash separately everything is fine. If you have a Microchip board, please try to run the sanitycheck for the Microchip board. I updated sanitycheck to create a hardware map for the Microchip board.

**Pull request merged**